From 93fd8e093c105902ffa0c522280020f8ad2df95f Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@manjaro.org>
Date: Thu, 13 May 2021 16:09:57 +0200
Subject: [PATCH 10/16] Bluetooth: btbcm: deassert bluetooth power gpio on
 shutdown

Some WiFi + Bluetooth modules manufactured with Broadcom chips experience
issues when reinitialized without cutting power to both the WiFi and
Bluetooth part of the module before reinitializing the module, as would be
the case during a reboot. This is the second half of a change ensuring
both the WiFi and Bluetooth section of such modules are powered down
during reboots.

Signed-off-by: Tobias Schramm <t.schramm@manjaro.org>
---
 drivers/bluetooth/hci_bcm.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/bluetooth/hci_bcm.c b/drivers/bluetooth/hci_bcm.c
index 8ea5ca8d71d6..eab8df59bd0f 100644
--- a/drivers/bluetooth/hci_bcm.c
+++ b/drivers/bluetooth/hci_bcm.c
@@ -1469,6 +1469,16 @@ static void bcm_serdev_remove(struct serdev_device *serdev)
 	hci_uart_unregister_device(&bcmdev->serdev_hu);
 }
 
+static void bcm_serdev_shutdown(struct serdev_device *serdev)
+{
+	int err;
+	struct bcm_device *bcmdev = serdev_device_get_drvdata(serdev);
+
+	err = bcm_gpio_set_power(bcmdev, false);
+	if (err)
+		dev_err(bcmdev->dev, "Failed to power down: %d\n", err);
+}
+
 #ifdef CONFIG_OF
 static struct bcm_device_data bcm4354_device_data = {
 	.no_early_set_baudrate = true,
@@ -1494,6 +1504,7 @@ MODULE_DEVICE_TABLE(of, bcm_bluetooth_of_match);
 static struct serdev_device_driver bcm_serdev_driver = {
 	.probe = bcm_serdev_probe,
 	.remove = bcm_serdev_remove,
+	.shutdown = bcm_serdev_shutdown,
 	.driver = {
 		.name = "hci_uart_bcm",
 		.of_match_table = of_match_ptr(bcm_bluetooth_of_match),
-- 
2.30.1

